pkg/reconciler/testing/factory.go

```diff
@@ -103,6 +110,75 @@ func MakeFactory(ctor Ctor, unstructured bool) Factory {
 	}
 }
 
+// injectionCtor functions create a k8s controller with given params.
+type injectionCtor func(context.Context, *Listers, configmap.Watcher) controller.Reconciler
+
+// MakeInjectionFactory creates a reconciler factory with fake clients and controller created by `ctor`.
+func MakeInjectionFactory(ctor injectionCtor, unstructured bool) Factory {
+	return func(t *testing.T, r *TableRow) (controller.Reconciler, ActionRecorderList, EventList, *FakeStatsReporter) {
+		ls := NewListers(r.Objects)
+
+		ctx := context.Background()
+		logger := logtesting.TestLogger(t)
+		ctx = logging.WithLogger(ctx, logger)
+
+		ctx, kubeClient := fakekubeclient.With(ctx, ls.GetKubeObjects()...)
+		ctx, client := fakeeventingclient.With(ctx, ls.GetEventingObjects()...)
+		ctx, dynamicClient := fakedynamicclient.With(ctx,
+			NewScheme(), ToUnstructured(t, r.Objects)...)
+
+		dynamicScheme := runtime.NewScheme()
+		for _, addTo := range clientSetSchemes {
+			addTo(dynamicScheme)
+		}
+
+		allObjects := ls.GetAllObjects()
+		if unstructured {
+			allObjects = ToUnstructured(t, allObjects)
+		}
+
+		// The dynamic client's support for patching is BS.  Implement it
+		// here via PrependReactor (this can be overridden below by the
+		// provided reactors).
+		dynamicClient.PrependReactor("patch", "*",
+			func(action ktesting.Action) (bool, runtime.Object, error) {
+				return true, nil, nil
+			})
+
+		eventRecorder := record.NewFakeRecorder(maxEventBufferSize)
+		ctx = controller.WithEventRecorder(ctx, eventRecorder)
+		statsReporter := &FakeStatsReporter{}
+		//ctx = reconciler.WithStatsReporter(ctx, statsReporter) // TODO: upstream stats interface from eventing to PKG
+
+		PrependGenerateNameReactor(&client.Fake)
+		PrependGenerateNameReactor(&dynamicClient.Fake)
+
+		// Set up our Controller from the fakes.
+		c := ctor(ctx, &ls, configmap.NewFixedWatcher())
+
+		for _, reactor := range r.WithReactors {
+			kubeClient.PrependReactor("*", "*", reactor)
+			client.PrependReactor("*", "*", reactor)
+			dynamicClient.PrependReactor("*", "*", reactor)
+		}
+
+		// Validate all Create operations through the serving client.
+		client.PrependReactor("create", "*", func(action ktesting.Action) (handled bool, ret runtime.Object, err error) {
+			// TODO(n3wscott): context.Background is the best we can do at the moment, but it should be set-able.
```

Comment:
context.TODO?

Reply:
Though, you already have ctx that you create above, so instead of creating another one here just use that?

Reply:
GOOD POINT!